WalletWasabi/WabiSabi/Client/WabiSabiApiClientWithDelay.cs

```diff
@@ -0,0 +1,90 @@
+		private ConcurrentStack<DateTimeOffset> InputRegistrationSchedule { get; }
+		private ConcurrentStack<DateTimeOffset> SignatureRequestsSchedule { get; }
+
+		public WabiSabiApiClientWithDelay(IWabiSabiApiRequestHandler innerClient,
+			int expectedNumberOfInputs,
+			TimeSpan inputRegistrationTimeFrame,
+			TimeSpan transactionSigningTimeFrame)
+		{
+			InnerClient = innerClient;
+			InputRegistrationSchedule = CreateSchedule(DateTimeOffset.Now, inputRegistrationTimeFrame, expectedNumberOfInputs);
+			SignatureRequestsSchedule = CreateSchedule(DateTimeOffset.Now, transactionSigningTimeFrame, expectedNumberOfInputs);
+		}
+
+		public IWabiSabiApiRequestHandler InnerClient { get; }
+
+		public async Task<InputRegistrationResponse> RegisterInputAsync(InputRegistrationRequest request, CancellationToken cancellationToken)
+		{
+			await DelayAccordingToScheduleAsync(InputRegistrationSchedule, cancellationToken).ConfigureAwait(false);
+			return await InnerClient.RegisterInputAsync(request, cancellationToken).ConfigureAwait(false);
+		}
+
+		public async Task<ConnectionConfirmationResponse> ConfirmConnectionAsync(ConnectionConfirmationRequest request, CancellationToken cancellationToken)
+		{
+			return await InnerClient.ConfirmConnectionAsync(request, cancellationToken).ConfigureAwait(false);
+		}
+
+		public async Task<ReissueCredentialResponse> ReissueCredentialAsync(ReissueCredentialRequest request, CancellationToken cancellationToken)
+		{
+			return await InnerClient.ReissueCredentialAsync(request, cancellationToken).ConfigureAwait(false);
+		}
+
+		public async Task<OutputRegistrationResponse> RegisterOutputAsync(OutputRegistrationRequest request, CancellationToken cancellationToken)
+		{
+			return await InnerClient.RegisterOutputAsync(request, cancellationToken);
+		}
+
+		public async Task RemoveInputAsync(InputsRemovalRequest request, CancellationToken cancellationToken)
+		{
+			await InnerClient.RemoveInputAsync(request, cancellationToken).ConfigureAwait(false);
+		}
+
+		public async Task SignTransactionAsync(TransactionSignaturesRequest request, CancellationToken cancellationToken)
+		{
+			await DelayAccordingToScheduleAsync(SignatureRequestsSchedule, cancellationToken).ConfigureAwait(false);
+			await InnerClient.SignTransactionAsync(request, cancellationToken).ConfigureAwait(false);
+		}
+
+		public async Task<RoundState[]> GetStatusAsync(CancellationToken cancellationToken)
+		{
+			return await InnerClient.GetStatusAsync(cancellationToken).ConfigureAwait(false);
+		}
+
+		private async Task DelayAccordingToScheduleAsync(ConcurrentStack<DateTimeOffset> schedule, CancellationToken cancellationToken)
+		{
+			if (schedule.TryPop(out var scheduledExecutionTime))
+			{
+				var timeToWait = scheduledExecutionTime - DateTimeOffset.Now;
+				var fixedTimeToWait = timeToWait < TimeSpan.Zero ? TimeSpan.Zero : timeToWait;
+
+				await Task.Delay(fixedTimeToWait, cancellationToken).ConfigureAwait(false);
+				return;
+			}
```

Comment:
For `BobClient`, assuming separate instances and that each one has its own delayed HTTP client internally this is overkill, as only one request will be made.

For `AliceClient` this will impose delays on requests that don't need them, such as the connection confirmation loop.

Perhaps a simpler way of addressing this is with a monotone, mutable `DateTimeOffset` that affects all requests, for `BobClient`s this would never be updated, and for `AliceClient`s this could be initialized to delay input registration, and then updated by the `CoinJoinClient` before signing phase?

That way each client is completely independent (instead of one shared `WabiSabiApiClientWithDelay`, each `AliceClient` or `BobClient` gets a separate instance) and the order in which offsets are popped off this shared queue wouldn't matter, which I think will make it easier to debug/follow as well as test.

Reply:
That's why i didn't implemented for any other calls, only for Register Inputs and Sing Transaction calls.

Anyway, I think there are only two ways to continue:

* Making this client a non-transparent one (i mean, currently it is a decorator) that can receive messages from the CJC or,
* Passing it a `RoundUpdater` instance what would allow this client to know from what point in time to start sampling time window.

I would prefer the second one.